miden-tx/src/host/tx_progress.rs

```diff
@@ -0,0 +1,163 @@
+    }
+
+    pub fn start_epilogue(&mut self, cycle: u32) {
+        self.epilogue.set_start(cycle);
+    }
+
+    pub fn end_epilogue(&mut self, cycle: u32) {
+        self.epilogue.set_end(cycle);
+    }
+
+    // DATA PRINT
+    // --------------------------------------------------------------------------------------------
+
+    /// Prints out the lengths of cycle intervals for each execution stage.
+    #[cfg(feature = "std")]
+    pub fn print_stages(&self) {
+        println!(
+            "Number of cycles it takes to execule:\n- Prologue: {},\n- Notes processing: {},",
+            self.prologue
+                .get_interval_len()
+                .map(|len| (len - SPAN_CREATION_SHIFT).to_string())
+                .unwrap_or("invalid interval".to_string()),
+            self.notes_processing
+                .get_interval_len()
+                .map(|len| (len - SPAN_CREATION_SHIFT).to_string())
+                .unwrap_or("invalid interval".to_string())
+        );
+
+        for (note_id, interval) in self.note_execution.iter() {
+            println!(
+                "--- Note {}: {}",
+                note_id.map(|id| id.to_hex()).unwrap_or("id_unavailable".to_string()),
+                interval
+                    .get_interval_len()
+                    .map(|len| (len - SPAN_CREATION_SHIFT).to_string())
+                    .unwrap_or("invalid interval".to_string())
+            )
+        }
+
+        println!(
+            "- Transaction script processing: {},\n- Epilogue: {}",
+            self.tx_script_processing
+                .get_interval_len()
+                .map(|len| (len - SPAN_CREATION_SHIFT).to_string())
+                .unwrap_or("invalid interval".to_string()),
+            self.epilogue
+                .get_interval_len()
+                .map(|len| (len - SPAN_CREATION_SHIFT).to_string())
+                .unwrap_or("invalid interval".to_string())
+        );
+    }
+}
+
+/// Stores the cycles corresponding to the start and the end of an interval.
+#[derive(Default)]
+struct CycleInterval {
+    start: Option<u32>,
+    end: Option<u32>,
+}
+
+impl CycleInterval {
+    pub fn new(start: u32) -> Self {
+        Self { start: Some(start), end: None }
+    }
+
+    pub fn set_start(&mut self, s: u32) {
+        self.start = Some(s);
+    }
+
+    pub fn set_end(&mut self, e: u32) {
+        self.end = Some(e);
+    }
+
+    /// Calculate the length of the interval
+    pub fn get_interval_len(&self) -> Option<u32> {
+        if let Some(start) = self.start {
+            if let Some(end) = self.end {
+                if end >= start {
+                    return Some(end - start);
+                }
+            }
+        }
+        None
+    }
+}
+
+// HELPER FUNCTIONS
+// ================================================================================================
+
+pub fn get_current_note_id<S: ProcessState>(process: &S) -> Result<Option<NoteId>, ExecutionError> {
```

Comment:
It feels like this function should be in a different module (I am not seeing it being used in this module).